tcslackbuildnotifier-core/src/main/java/slacknotifications/SlackNotificationImpl.java

```diff
@@ -239,50 +241,100 @@ private void postViaWebHook() throws IOException {
             Loggers.SERVER.info("SlackNotificationListener :: Preparing message for URL " + url);
 
             WebHookPayload requestBody = new WebHookPayload();
-            
-            requestBody.setChannel(this.getChannel());
-            requestBody.setUsername(this.getBotName());
-            requestBody.setIcon_url(this.getIconUrl());
 
-            HttpPost httppost = new HttpPost(url);
+            List<Commit> commits = this.payload.getCommits();
 
-            if (this.payload != null) {
-                requestBody.setText(payload.getBuildDescriptionWithLinkSyntax());
-                requestBody.setAttachments(getAttachments());
+            List<String> usersSentTo = new ArrayList<String>();
+            boolean sendToUser;
+
+            //send to default channel if enabled
+            if(sendDefaultChannel){
+                sendToChannel(this.getChannel(), requestBody, url);
             }
 
-            String bodyParam = String.format("payload=%s", URLEncoder.encode(requestBody.toJson(), UTF8));
+            if(sendUsers){
+                String slackUser;
+                for(Commit commit : commits){
+                    if(commit.hasSlackUserId()){
+                        sendToUser = true;
+                        slackUser = commit.getSlackUserId();
+
+                        /*Skip sending to the user if the commit was a merge from someone else
+                        * or if they were already sent the message*/
+                        for(String user : usersSentTo){
+                            if(user.equals(slackUser)){
+                                sendToUser = false;
+                                break;
+                            }
+                        }
+                        if(sendToUser){
+                            if(commit.getDescription().contains("Merge pull request") && !commit.getDescription().contains(commit.getUserName())){
+                                sendToUser = false;
+                            }
+                        }
+
+                        if(sendToUser){
+                            if(slackUser.charAt(0) != '@'){
+                                slackUser = "@" + slackUser;
+                            }
+                            try {
+                                usersSentTo.add(slackUser.substring(1));
+                                sendToChannel(slackUser, requestBody, url);
+                            }
+                            catch(IOException e){
+                                //failed to find slack user ID
+                            }
+                        }
+                    }
```

Comment:
Suggested way to shorten this for loop by reorganizing a little:
```
String slackUser = commit.getSlackUserId();
if (!slackUser.startsWith("@")) {
  slackUser = "@" + slackUser;
}

if (!usersSentTo.contains(slackUser) && !isMergeCommit(commit)) {
  try {
    sendToChannel(slackUser, requestBody, url);
    usersSentTo.add(slackUser);
  } catch(IOException e) {
    // TODO Better to log this than just a comment
  }
}
```

Implement the new method `isMergeCommit`.
Notice, I moved the operation to add a slack user to the `usersSentTo` set after attempting to send the message. Otherwise the set might indicate a user got a message sent, even if there was an IOException.

Reply:
I suggest the new method `isMergeCommit` because that will make the code more readable. The if statement if statement is a bit long otherwise and harder to read.